proposals/xxxx-die.md

```diff
@@ -0,0 +1,208 @@
+  
+  /// Die because this method or function has not yet been implemented.
+  public static let notImplemented = FatalReason("Not yet implemented.")
+  
+  /// Die because a default method must be overriden by a
+  /// subtype or extension.
+  public static let mustOverride = FatalReason("Must be overriden in subtype.")
+  
+  /// Die because this functionality should never be called,
+  /// typically to silence requirements.
+  public static let uncallable = FatalReason("Should never be called.")
+  
+  /// An underlying string-based cause for a fatal error.
+  public let reason: String
+  
+  /// Establishes a new instance of a `FatalReason` with a string-based explanation.
+  public init(_ reason: String) {
+    self.reason = reason
+  }
+  
+  /// Conforms to CustomStringConvertible, allowing reason to 
+  /// print directly to complaint.
+  public var description: String {
+    return reason
+  }
+}
+
+/// Unconditionally prints a given message and stops execution.
+///
+/// - Parameters:
+///   - reason: A predefined `FatalReason`.
+///   - function: The name of the calling function to print with `message`. The 
+///     default is the calling scope where `fatalError(because:, function:, file:, line:)`
+///     is called.
+///   - file: The file name to print with `message`. The default is the file
+///     where `fatalError(because:, function:, file:, line:)` is called.
+///   - line: The line number to print along with `message`. The default is the
+///     line number where `fatalError(because:, function:, file:, line:)` is called.
+@inlinable // FIXME(sil-serialize-all)
+@_transparent
+public func fatalError(
+  because reason: FatalReason, 
+  function: StaticString = #function,
+  file: StaticString = #file, 
+  line: UInt = #line
+) -> Never {
+  fatalError("\(function): \(reason)", file: file, line: line)
+}
+```
+
+You call `fatalError(because:)` using the static entry points:
+
+```swift
+fatalError(because: .unreachable)
+fatalError(because: .notImplemented)
+fatalError(because: .mustOverride)
+fatalError(because: .uncallable)
+```
+
+A few points about this design:
+
+* The contextual members (`function`, `file`, and `line`) are automatically picked up by the defaulted properties and passed to `fatalError`.
+* The `FatalReason` namespacing allows each reason to be auto-completed by a hosting IDE.
+* The text is intentionally developer-facing and unlocalized, as with other warning and error content.
+* The `FatalReason` namespace encourages additions specific to fatal outcomes. As its name suggests, it is not a shared repository for developer-facing strings.
+
+This design allows you to extend `FatalReason` to introduce custom causes of programmatic run-time failure common to your in-house work, such as:
+
+```swift
+// Custom outlet complaint
+extension FatalReason {
+  static func missingOutlet(_ name: String) -> Reason {
+    return Reason("Outlet \"\(name)\" is not connected")
```

Comment:
You probably want to return and initialize a `FatalReason` here.

Reply:
Fixed, thank you!